synapse/handlers/presence.py

```diff
@@ -967,27 +988,55 @@ async def user_syncing(
             self.user_to_num_current_syncs[user_id] = curr_sync + 1
 
             prev_state = await self.current_state_for_user(user_id)
-            if prev_state.state == PresenceState.OFFLINE:
-                # If they're currently offline then bring them online, otherwise
-                # just update the last sync times.
+            # If they're busy then they don't stop being busy just by syncing,
+            # so just update the last sync time.
+            if prev_state.state == PresenceState.BUSY:
                 await self._update_states(
                     [
                         prev_state.copy_and_replace(
-                            state=PresenceState.ONLINE,
+                            last_user_sync_ts=self.clock.time_msec(),
+                        )
+                    ]
+                )
+            # If they were offline, then bring them online (update the presence
+            # state to the new state and update the last_active_ts)
+            elif prev_state.state == PresenceState.OFFLINE:
+                await self._update_states(
+                    [
+                        prev_state.copy_and_replace(
+                            # We do want to update state here but currently do so with separate
+                            # set_state below.
+                            # state=presence_state,
                             last_active_ts=self.clock.time_msec(),
                             last_user_sync_ts=self.clock.time_msec(),
                         )
                     ]
                 )
+            # otherwise, set the new presence state & update the last sync time,
+            # but don't update last_active_ts as this isn't an indication that
+            # they've been active.
             else:
                 await self._update_states(
                     [
                         prev_state.copy_and_replace(
-                            last_user_sync_ts=self.clock.time_msec()
+                            # We do want to update state here but currently do so with separate
+                            # set_state below.
+                            # state=presence_state,
+                            last_user_sync_ts=self.clock.time_msec(),
                         )
                     ]
                 )
```

Comment:
I'm finding this very confusing (sorry!). Could we do the following refactoring:
1. Move the `if prev_state.state != PresenceState.BUSY:` block before all this, but before `await self.current_state_for_user`. That way it matches the order of things in worker mode. It's not so much the order that matters, but the value of `prev_state.state` when the `if prev_state.state == PresenceState.OFFLINE` block gets run.
2. Roll back this block of changes to bring `user_syncing` and `update_external_syncs_row` back into line.
3. Replace `prev_state.state` with the presence after the `set_state` call, because in worker mode, we'd be reading the presence after the call to `set_state`.

ie.
```py
            prev_state = await self.current_state_for_user(user_id)

            if prev_state.state != PresenceState.BUSY:
                ...
                await self.set_state(
                    UserID.from_string(user_id), {"presence": presence_state}, True
                )
                prev_state = prev_state.copy_and_replace(state=presence_state)

            if prev_state.state == PresenceState.OFFLINE:
                # If they're currently offline then bring them online, otherwise
                # just update the last sync times.
                await self._update_states(
                    [
                        prev_state.copy_and_replace(
                            state=PresenceState.ONLINE,
                            last_active_ts=self.clock.time_msec(),
                            last_user_sync_ts=self.clock.time_msec(),
                        )
                    ]
                )
            else:
                await self._update_states(
                    [
                        prev_state.copy_and_replace(
                            last_user_sync_ts=self.clock.time_msec()
                        )
                    ]
                )
```

which is analogous to the following code path in worker mode:
```py
        prev_state = await self.current_state_for_user(user_id)
        if prev_state != PresenceState.BUSY:
            await self.set_state(
                UserID.from_string(user_id), {"presence": presence_state}, True
            )

        # followed by a call to `update_external_syncs_row`
```

Reply:
Yeah, I think I see how that makes more sense - done

Reply:
...although this was then re-checking the value of `prev_state.state` after modifying it, so I've made it make a copy in a new variable before making modifications

Reply:
> ...although this was then re-checking the value of prev_state.state after modifying it

That part was intentional: in worker mode we call `set_state` on the worker, then bounce to the writer, which reads a `prev_state` that comes from the `set_state` call.

I hope all the back and forth of review isn't proving too bad. I think we're almost there for what it's worth!

Reply:
Hmm, I see what you mean. Unfortunately doing that caused a bunch of sytest failures: https://github.com/matrix-org/synapse/runs/5973187907?check_suite_focus=true - I thought the CI sytest ran with the synapse in worker mode, but I guess not if it's executing this code path? Does this suggest that the worker mode code would fail this sytest?

Reply:
That's unfortunate. CI has multiple sytest runs, including both single process and worker configurations. The worker mode configuration seems to be passing. I'll try to investigate what's going on tomorrow.

Reply:
Turns out `prev_state = prev_state.copy_and_replace(state=presence_state)` is insufficient, since it doesn't pick up the update to `last_active_ts` from the `set_state` call.

Instead, `prev_state = await self.current_state_for_user(user_id)` after `set_state` _ought_ to make sytest pass.
This looks awful, but ought to hit the in-memory cache, since we've just written to it a few lines before.

Reply:
Ah, makes sense - done

Reply:
That's done the trick!